src/NHibernate/Proxy/NHibernateProxyBuilder.cs

```diff
@@ -0,0 +1,366 @@
+			// this works, but the generated assembly will fail PeVerify (cannot use in medium trust for example)
+			if (baseConstructor == null || baseConstructor.IsPrivate || baseConstructor.IsAssembly)
+				baseConstructor = ProxyFactory.defaultBaseConstructor;
+
+			var IL = constructor.GetILGenerator();
+
+			constructor.SetImplementationFlags(MethodImplAttributes.IL | MethodImplAttributes.Managed);
+
+			IL.Emit(OpCodes.Ldarg_0);
+			IL.Emit(OpCodes.Call, baseConstructor);
+
+			IL.Emit(OpCodes.Ldarg_0);
+			IL.Emit(OpCodes.Ldarg_1);
+			IL.Emit(OpCodes.Stfld, lazyInitializerField);
+
+			IL.Emit(OpCodes.Ldarg_0);
+			IL.Emit(OpCodes.Ldarg_2);
+			IL.Emit(OpCodes.Stfld, proxyFactoryField);
+
+			IL.Emit(OpCodes.Ret);
+
+			return constructor;
+		}
+
+		static void ImplementDeserializationConstructor(TypeBuilder typeBuilder)
+		{
+			var parameterTypes = new[] {typeof (SerializationInfo), typeof (StreamingContext)};
+			var constructor = typeBuilder.DefineConstructor(constructorAttributes, CallingConventions.Standard, parameterTypes);
+			constructor.SetImplementationFlags(MethodImplAttributes.IL | MethodImplAttributes.Managed);
+
+			var IL = constructor.GetILGenerator();
+			//Everything is done in NHibernateProxyObjectReference, so just return data.
+			IL.Emit(OpCodes.Ret);
+		}
+
+		static void ImplementGetObjectData(System.Type baseType, IReadOnlyCollection<System.Type> baseInterfaces, TypeBuilder typeBuilder, FieldBuilder lazyInitializerField, FieldBuilder proxyFactoryField)
+		{
+			const MethodAttributes attributes = MethodAttributes.Public | MethodAttributes.HideBySig |
+			                                    MethodAttributes.Virtual;
+			var parameterTypes = new[] {typeof (SerializationInfo), typeof (StreamingContext)};
+
+			MethodBuilder methodBuilder =
+				typeBuilder.DefineMethod("GetObjectData", attributes, typeof (void), parameterTypes);
+
+			ILGenerator IL = methodBuilder.GetILGenerator();
+			//LocalBuilder proxyBaseType = IL.DeclareLocal(typeof(Type));
+
+			// info.SetType(typeof(NHibernateProxyObjectReference));
+			IL.Emit(OpCodes.Ldarg_1);
+			IL.Emit(OpCodes.Ldtoken, typeof (NHibernateProxyObjectReference));
+			IL.Emit(OpCodes.Call, ReflectionCache.TypeMethods.GetTypeFromHandle);
+			IL.Emit(OpCodes.Callvirt, ProxyFactory.setType);
+
+			//this.__proxyFactory
+			IL.Emit(OpCodes.Ldarg_0);
+			IL.Emit(OpCodes.Ldfld, proxyFactoryField);
+
+			//this.LazyInitializer.Identifier
+			EmitGetLazyInitializer(IL);
+			IL.Emit(OpCodes.Callvirt, LazyInitializerIdentifierProperty.GetMethod);
+
+			var constructor = typeof(NHibernateProxyObjectReference).GetConstructor(
+				new[]
+				{
+					typeof(IProxyFactory),
+					typeof(object),
+				});
+			IL.Emit(OpCodes.Newobj, constructor);
+
+			IL.Emit(OpCodes.Ldarg_1);
+			IL.Emit(OpCodes.Ldarg_2);
+
+			IL.Emit(OpCodes.Callvirt, typeof(ISerializable).GetMethod(nameof(ISerializable.GetObjectData)));
```

Comment:
So we have now two different ways of implementing the an `IObjectReference` with proxies, since the one for `DynamicProxy.ProxyFactory` does not delegate the `GetObjectData` to `DynamicProxy.ProxyObjectReference`. I prefer the pattern you have done for `NHibernateProxyObjectReference`, but then we should align `ProxyObjectReference` on the same pattern.

Reply:
Not going to bother with DynamicProxy namespace. Going to obsolete it after converting lazy properties proxies to the same approach.